### PR TITLE
Extend aot autograd dedup guards to params, stop using positions

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96774

The purpose of this PR is to remove reliance on argument positions in dedup guards, AND extend the functionality to params.

A version of this PR was stamped prior https://github.com/pytorch/pytorch/pull/95831 - but was kinda gross, because it was based on an underlying PR that did way too much with source names. 

This PR leaves most of that alone, in favor of just reusing the same name standardization logic that dynamo module registration does. 

cc @soumith @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_functorch/aot_autograd.py [hunk 1]
```diff
@@ -2851,9 +2846,54 @@ def functional_call(*args, **kwargs):
...
+                # Reconstruct this to ensure we get the same name contract as dynamo
+                new_source = AttrSource(base, param_name)
+                # Sanitize the name, just as dynamo does
+                new_name = new_source.name()
+                # BEGIN DUPLICATIVE LOGIC TO DEDUP
+                new_name = re.sub(r"\[(\d+)\]", r"_\g<1>", new_name)
+                new_name = re.sub(r"[^a-zA-Z0-9]", "_", new_name)
+                if not new_name[0].isalpha():
+                    new_name = "sub" + new_name
+                # END DUPLICATIVE LOGIC TO DEDUP
```
Comment (voznesenskym) [inline torch/_functorch/aot_autograd.py:2881]: I am leaving it in this PR as such, for now, to make it easier to understand.  The 3 lines of duplicate code could ostensibly be moved to a util, but it is not overly important.

File: torch/_dynamo/output_graph.py [hunk 2]
```diff
@@ -458,7 +473,29 @@ def wrap_name(module_key):
...
         for i in itertools.count():
             if name not in self.nn_modules:
+                assert self.nn_modules_sources is not None
+                assert name not in self.nn_modules_sources
                 self.nn_modules[name] = target
+                self.nn_modules_sources[name] = source
+                if isinstance(target, torch.nn.Module):
+                    # annoying, but there are cases when we do not have parameters
+                    # see test_nn_moduledict_contains
+                    def reg(n, p):
```
Comment (wconstab) [inline torch/_dynamo/output_graph.py:483]: nit, use leaf_name and param_or_buffer?

**Review by wconstab — COMMENTED** [review]
i think this lgtm. several questions inline.  nice work!

File: torch/_functorch/aot_autograd.py [hunk 3]
```diff
@@ -2851,9 +2846,54 @@ def functional_call(*args, **kwargs):
...
+        for name in params.keys():
+            if name in mod._nn_module_sources:
+                # Easy case
+                aot_autograd_arg_pos_to_source.append(mod._nn_module_sources[name])
+            else:
+                # Complex case - an attribute is found with "." in its name!
+                # This is not valid for dynamo named params, so, we need to reconstruct
+                # the source.
+                # A name is always a base_name, and a param_name
+                base_name, param_name = name.split(".")
```
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_functorch/aot_autograd.py:2867]: this comes from named_params and named_buffers, which produce a f"{module}.{name}" syntax, that we need to turn into the full proper name.

File: torch/_functorch/aot_autograd.py [hunk 4]
```diff
@@ -1800,22 +1802,15 @@ def add_dupe_args(args):
     deduped_flat_args = remove_dupe_args(flat_args)
 
     tracing_context = TracingContext.get()
-    if tracing_context:
+    if tracing_context and aot_config.aot_autograd_arg_pos_to_source and len(aot_config.aot_autograd_arg_pos_to_source) > 0:
```
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_functorch/aot_autograd.py:1805]: No - because this is valid for non dynamo entry points, which want dedup but do not have guard systems.

File: torch/_dynamo/guards.py [hunk 5]
```diff
@@ -699,19 +699,9 @@ def compile_check_fn(
...
-                assert self.output_graph.graphargs[
-                    pos_a
-                ].is_tensor, "Deduped arg must be a tensor"
-                assert self.output_graph.graphargs[
-                    pos_b
-                ].is_tensor, "Deduped arg must be a tensor"
-                code_part = f"{self.output_graph.graphargs[pos_a].source.name()} is {self.output_graph.graphargs[pos_b].source.name()}"  # noqa: B950
+                source_a = guard.input_source_a
+                source_b = guard.input_source_b
+                code_part = f"{source_a.name()} is {source_b.name()}"
```
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_dynamo/guards.py:704]: :)

File: torch/_dynamo/output_graph.py [hunk 6]
```diff
@@ -391,7 +397,12 @@ def register_attr_or_module(
         options["guards"] = set(options.get("guards", []))
         assert "source" in options
         source = options["source"]
-        if isinstance(target, torch.Tensor):
+        if isinstance(source, ParamBufferSource):
```
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_dynamo/output_graph.py:400]: yea, good call

File: torch/_dynamo/output_graph.py (hunk 2 above)
Comment (voznesenskym) [reply torch/_dynamo/output_graph.py:483]: sure :)

File: torch/_dynamo/variables/torch.py [hunk 7]
```diff
@@ -873,7 +874,13 @@ def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
...
             tx.restore_graphstate(checkpoint)
 
-            return output, graph, guards, nn_modules, comparable_state
+            return (
+                output,
+                graph,
+                guards,
+                nn_modules,
+                comparable_state,
+            )
```
Comment (ezyang) [inline torch/_dynamo/variables/torch.py:883]: no change

File: torch/_functorch/aot_autograd.py (hunk 4 above)
Comment (ezyang) [inline torch/_functorch/aot_autograd.py:1805]: nit: an empty list will interpret as bool false. So you do not need the length test.

File: torch/_functorch/aot_autograd.py [hunk 8]
```diff
@@ -2851,9 +2846,54 @@ def functional_call(*args, **kwargs):
...
+                aot_autograd_arg_pos_to_source.append(mod._nn_module_sources[new_name])
+
+
+    # Next, the input args
     full_args.extend(args)
 
+    # And now their dynamo equivalent, if there
+    if hasattr(mod, "graph"):
+        arg_srcs = [x._dynamo_source for x in mod.graph.nodes if x.op == "placeholder" and hasattr(x, "_dynamo_source")]
+        aot_autograd_arg_pos_to_source.extend(arg_srcs)
```
Comment (ezyang) [inline torch/_functorch/aot_autograd.py:2892]: If the input module is not a GraphModule, then aot_autograd_arg_pos_to_source is invalid and we ought to pass None to AOTConfig

File: torch/_functorch/aot_autograd.py [hunk 9]
```diff
@@ -2851,9 +2846,54 @@ def functional_call(*args, **kwargs):
         bw_compiler = fw_compiler
 
     full_args = []
+    # First, the params
     full_args.extend(params_flat)
+
+    aot_autograd_arg_pos_to_source = []
+    # Then, the params 1:1 mapped sources, if relevant.
+    if hasattr(mod, "_nn_module_sources"):
```
Comment (ezyang) [inline torch/_functorch/aot_autograd.py:2854]: This isn't the source of an nn module? It's the source of a parameter no?

File: torch/_dynamo/output_graph.py [hunk 10]
```diff
@@ -458,7 +473,29 @@ def wrap_name(module_key):
         base = name
         for i in itertools.count():
             if name not in self.nn_modules:
+                assert self.nn_modules_sources is not None
+                assert name not in self.nn_modules_sources
                 self.nn_modules[name] = target
+                self.nn_modules_sources[name] = source
+                if isinstance(target, torch.nn.Module):
+                    # annoying, but there are cases when we do not have parameters
+                    # see test_nn_moduledict_contains
```
Comment (ezyang) [inline torch/_dynamo/output_graph.py:482]: I don't understand this comment

**Comment by ezyang** [comment]
This looks a lot better! There's a lot of fuzz in https://github.com/pytorch/pytorch/pull/96774/files/d32a4df212f3eeb2ca14b6e11e8a0679efbf4566#diff-1bbe80a42da83f0b0e1386c3aca92323930626a65173c7d7e838931446fb4324 which is surprising to me. Naively, what I would have expected is that when Dynamo attaches parameters/buffers onto the GraphModule, it can work out what the sources for them should be. This does not appear to be what the code is doing, and I am not sure why it's doing it this way.

File: torch/_functorch/aot_autograd.py (hunk 1 above)
Comment (wconstab) [reply torch/_functorch/aot_autograd.py:2881]: why not do the util now actually?  seems equally hard to write the explanation and todo haha

File: torch/_dynamo/output_graph.py (hunk 10 above)
Comment (voznesenskym) [reply torch/_dynamo/output_graph.py:482]:
```
 class M(torch.nn.Module):
            def __init__(self, module_dict):
                super().__init__()
                self.module_dict = module_dict

            def forward(self, x):
                if "foo" in self.module_dict:
                    x = torch.mul(x, 1.0)
                x = torch.add(x, 1.0)
                return x

        module_dict = torch.nn.ModuleDict({"foo": torch.nn.Conv2d(1, 1, 1)})
        m = M(module_dict)
```

Is valid, but breaks some behavior and leads to a None `._parameters`

File: torch/_functorch/aot_autograd.py (hunk 9 above)
Comment (voznesenskym) [reply torch/_functorch/aot_autograd.py:2854]: Top level params would just have a name like `weight` or `buf` or whatever. It is a small case.

File: torch/_functorch/aot_autograd.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_functorch/aot_autograd.py:2867]: It does. Thats why we are able to read them out _nn_module_sources. Dynamo absolutely got to it.

File: torch/_functorch/aot_autograd.py (hunk 8 above)
Comment (voznesenskym) [reply torch/_functorch/aot_autograd.py:2892]: Sg

File: torch/_functorch/aot_autograd.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_functorch/aot_autograd.py:2867]: Let me explain this a little more:

1) Dynamo iterates over params, and buffers.
2) Dynamo writes them out as a full path, with all `.` and `[]` removed : `some_full_path__0__to_a_module_weights`
3) Dynamo stores a source at `some_full_path__0__to_a_module_weights` that points to `some.full_path[0].to.a.module.weights` in user code.
4) Dynamo stores modules like `some_full_path__0__to_a` on itself, in `output_graph.nn_modules`. This means that `gm.some_full_path__0__to_a` points to a valid module. 
5) In AOTAutograd, we iterate over the passed in gm's params and buffers. They have names like `some_full_path__0__to_a.weight`, because the `gm` came from dynamo, but the buffers and params are still homed on their respective modules, even though all the seen modules got floated up to top level and flattened. 
6) In order to match the AOTAutograd seen param name, we need to re-do the normalization via an AttrSource and parallel name change, same as dynamo, to turn `some_full_path__0__to_a.weight` into `some_full_path__0__to_a_module_weights`, the same key that dynamo saw in (2).

**Comment by voznesenskym** [comment]
> This looks a lot better! There's a lot of fuzz in https://github.com/pytorch/pytorch/pull/96774/files/d32a4df212f3eeb2ca14b6e11e8a0679efbf4566#diff-1bbe80a42da83f0b0e1386c3aca92323930626a65173c7d7e838931446fb4324 which is surprising to me. Naively, what I would have expected is that when Dynamo attaches parameters/buffers onto the GraphModule, it can work out what the sources for them should be. This does not appear to be what the code is doing, and I am not sure why it's doing it this way.

See my comment above.

File: torch/_functorch/aot_autograd.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_functorch/aot_autograd.py:2867]: An alternative approach would be to just move all these params and buffers to top level, and give them their full name like `some_full_path__0__to_a_module_weights`, however, that feels like it would change the actual model structure, which this PR explicitly avoids because I think that makes it way more complicated.

File: torch/_functorch/aot_autograd.py (hunk 9 above)
Comment (ezyang) [reply torch/_functorch/aot_autograd.py:2854]: It is confusing for both parameters and modules to be in a map called `_nn_module_sources`!

File: torch/_functorch/aot_autograd.py (hunk 4 above)
Comment (voznesenskym) [reply torch/_functorch/aot_autograd.py:1804]: This can be None

File: torch/_inductor/overrides.py [hunk 11]
```diff
@@ -110,22 +110,23 @@ def fetch_attr(target: str, mod):
...
+# Note - This is a CPU only pass, see fuse_fx
 def remove_identity(gm: torch.fx.GraphModule):
     """
     Removes all identity layers from the module.
     """
 
     class IdentityRemover(torch.fx.Transformer):
         def call_module(self, target, args, kwargs):
-            if isinstance(self.submodules[target], nn.Identity):
+            if isinstance(self.module._modules[target], nn.Identity):
```
Comment (ezyang) [inline torch/_inductor/overrides.py:121]: This looks like trouble, did you make a BC breaking change?

File: test/export/test_export.py [hunk 12]
```diff
@@ -55,6 +55,7 @@ def forward(self, x):
         self.assertEqual(exported_program.fw_module(*inp)[0], mod(*inp))
 
     @unittest.skipIf(not torchdynamo.is_dynamo_supported(), "dynamo doesn't support")
+    @unittest.skipIf(True, "Test failing unclear internal invariant for do_not_use_experimental_export")
```
Comment (ezyang) [inline test/export/test_export.py:58]: @SherlockNoMad can you help take a look?

@voznesenskym what's the failure?

File: torch/_dynamo/output_graph.py [hunk 13]
```diff
@@ -446,11 +456,6 @@ def wrap_name(module_key):
                 )
 
         assert self.nn_modules is not None
-        for k, v in self.nn_modules.items():
-            if v is target:
-                # it already exists
-                return wrap_name(k)
```
Comment (ezyang) [inline torch/_dynamo/output_graph.py:452]: What's going on here?

File: torch/_inductor/overrides.py (hunk 11 above)
Comment (voznesenskym) [reply torch/_inductor/overrides.py:121]: I do not think so! For some reason, the name is not in self.submodules, but it is in self.module._modules! I need to look at this closer.

File: torch/_dynamo/output_graph.py (hunk 10 above)
Comment (ezyang) [reply torch/_dynamo/output_graph.py:481]: @albanD @jbschlosser do you know about this? It seems weird to me that there exist modules for which `named_parameter` will fail if you call it on them.

File: torch/_inductor/overrides.py (hunk 11 above)
Comment (voznesenskym) [reply torch/_inductor/overrides.py:121]: I think I know what this is though, let me double check...

File: torch/_functorch/aot_autograd.py [hunk 14]
```diff
@@ -2988,9 +2982,33 @@ def functional_call(*args, **kwargs):
...
     full_args.extend(args)
 
+    # And now their dynamo equivalent, if there
+    if hasattr(mod, "graph"):
+        if aot_autograd_arg_pos_to_source is None:
+            aot_autograd_arg_pos_to_source = []
+        arg_srcs = [x._dynamo_source for x in mod.graph.nodes if x.op == "placeholder" and hasattr(x, "_dynamo_source")]
+        aot_autograd_arg_pos_to_source.extend(arg_srcs)
+
+    if aot_autograd_arg_pos_to_source:
```
Comment (ezyang) [inline torch/_functorch/aot_autograd.py:3009]: `aot_autograd_arg_pos_to_source is None`

If it's empty, then I DO want to know about mismatch

**Review by ezyang — APPROVED** [review]
Much better, thanks.

File: torch/_functorch/aot_autograd.py [hunk 15]
```diff
@@ -2988,9 +2982,33 @@ def functional_call(*args, **kwargs):
...
+        assert len(full_args) == len(aot_autograd_arg_pos_to_source)
```
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_functorch/aot_autograd.py:3010]: I am not in a rush, let me see if I can make this fail earlier. 

We should still keep this check here, for safety, but I agree that sooner is better!

[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_functorch/aot_autograd.py:3013]: If you're going to refactor this, the other problem is that you implicitly have an invariant that if `_param_name_to_source`, then `_dynamo_source` must be specified on all placeholders, but then your code is written as if it's possible for them to be out of sync:

```
                    # ... but not here!
                    if aot_autograd_arg_pos_to_source is None:
                        aot_autograd_arg_pos_to_source = []
```

If `_dynamo_source` is set, `_param_name_to_source` must have been set, and you should be asserting accordingly! In other words, opt for performing the STRONGEST ASSERT you can prove to yourself should hold, it is more useful than a weak assert.

[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_functorch/aot_autograd.py:3013]: > _dynamo_source is set, _param_name_to_source

You could have a graph with no params, right?

[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_functorch/aot_autograd.py:3013]: Then param name to source should be set as an empty dict.
